spec.bs

```diff
@@ -243,8 +250,200 @@ Each {{URLPattern}} object has an associated <dfn for=URLPattern>hash component<
 </div>
 
+<div algorithm>
+  The <dfn method for="URLPattern">test(|input|, |baseURL|)</dfn> method steps are:
+
+  1. Let |result| be the result of [=match=] given [=this=], |input|, and |baseURL|.
+  1. If |result| is null, return false.
+  1. Return true.
+</div>
+
+<div algorithm>
+  The <dfn method for="URLPattern">exec(|input|, |baseURL|)</dfn> method steps are:
+
+  1. Return the result of [=match=] given [=this=], |input|, and |baseURL|.
+</div>
+
+<div algorithm>
+  To perform a <dfn>match</dfn> given a {{URLPattern}} |urlpattern|, a {{URLPatternInput}} |input|, and an optional string |baseURLString|:
+
+  1. Let |protocol| be the empty string.
+  1. Let |username| be the empty string.
+  1. Let |password| be the empty string.
+  1. Let |hostname| be the empty string.
+  1. Let |port| be the empty string.
+  1. Let |pathname| be the empty string.
+  1. Let |search| be the empty string.
+  1. Let |hash| be the empty string.
+  1. Let |inputs| be an empty array.
+  1. Push |input| on to |inputs|.
+  1. If |input| is a {{URLPatternInit}} then:
+    1. If |baseURLString| is not null, throw a TypeError.
+    1. Let applyResult be the result of [=process a URLPatternInit=] given |input|, "url", |protocol|, |username|, |password|, |hostname|, |port|, |pathname|, |search|, and |hash|. If this throws an exception, catch it, and return null.
+    1. Set |protocol| to |applyResult|'s {{URLPatternInit/protocol}}.
+    1. Set |username| to |applyResult|'s {{URLPatternInit/username}}.
+    1. Set |password| to |applyResult|'s {{URLPatternInit/password}}.
+    1. Set |hostname| to |applyResult|'s {{URLPatternInit/hostname}}.
+    1. Set |port| to |applyResult|'s {{URLPatternInit/port}}.
+    1. Set |pathname| to |applyResult|'s {{URLPatternInit/pathname}}.
+    1. Set |search| to |applyResult|'s {{URLPatternInit/search}}.
+    1. Set |hash| to |applyResult|'s {{URLPatternInit/hash}}.
+  1. Else:
+    1. Let |baseURL| be null.
+    1. If |baseURLString| is not null then:
+      1. Set |baseURL| to the result of [=URL parser|parsing=] |baseURLString|.
+      1. If |baseURL| is failure, return null.
+      1. Push |baseURL| on to |inputs|.
+    1. Let |url| be the result of [=URL parser|parsing=] |input| given |baseURL|.
+      1. If |url| is failure, return null.
+    1. Set |protocol| to |url|'s {{URL/protocol}}.
+    1. Set |username| to |url|'s {{URL/username}}.
+    1. Set |password| to |url|'s {{URL/password}}.
+    1. Set |hostname| to |url|'s {{URL/hostname}}.
+    1. Set |port| to |url|'s {{URL/port}}.
+    1. Set |pathname| to |url|'s {{URL/pathname}}.
+    1. Set |search| to |url|'s {{URL/search}}.
+    1. Set |hash| to |url|'s {{URL/hash}}.
+  1. Let |protocolExecResult| be the result of [=RegExpExec=] given |urlpattern|'s [=URLPattern/protocol component=]'s [=component/regular expression=] and |protocol|.
+  1. If |protocolExecResult| is null then return null.
```

Comment:
Something that might make the spec easier to read is if you grouped all these steps at the end. `If |protocolExecResult|, |usernameExecResult|, |passwordExecResult|, ..., or |hashExecResult| is null, then return null`.